// @clerk/clerk-expo/dist/index.js
"use strict";
exports.__esModule = true;
exports.Clerk = exports.MagicLinkErrorCode = exports.isKnownError = exports.isMetamaskError = exports.isMagicLinkError = exports.isClerkAPIResponseError = exports.WithUser = exports.withUser = exports.WithSession = exports.withSession = exports.WithClerk = exports.withClerk = exports.useUser = exports.useSignUp = exports.useSignIn = exports.useSessionList = exports.useSession = exports.useOrganizations = exports.useOrganizationList = exports.useOrganization = exports.useMagicLink = exports.useClerk = exports.useAuth = exports.SignedOut = exports.SignedIn = exports.MultisessionAppSupport = exports.ClerkLoading = exports.ClerkLoaded = void 0;
var tslib_1 = require("tslib");
var clerk_react_1 = require("@clerk/clerk-react");
tslib_1.__createBinding(exports, clerk_react_1, "ClerkLoaded");
tslib_1.__createBinding(exports, clerk_react_1, "ClerkLoading");
tslib_1.__createBinding(exports, clerk_react_1, "MultisessionAppSupport");
tslib_1.__createBinding(exports, clerk_react_1, "SignedIn");
tslib_1.__createBinding(exports, clerk_react_1, "SignedOut");
tslib_1.__createBinding(exports, clerk_react_1, "useAuth");
tslib_1.__createBinding(exports, clerk_react_1, "useClerk");
tslib_1.__createBinding(exports, clerk_react_1, "useMagicLink");
tslib_1.__createBinding(exports, clerk_react_1, "useOrganization");
tslib_1.__createBinding(exports, clerk_react_1, "useOrganizationList");
tslib_1.__createBinding(exports, clerk_react_1, "useOrganizations");
tslib_1.__createBinding(exports, clerk_react_1, "useSession");
tslib_1.__createBinding(exports, clerk_react_1, "useSessionList");
tslib_1.__createBinding(exports, clerk_react_1, "useSignIn");
tslib_1.__createBinding(exports, clerk_react_1, "useSignUp");
tslib_1.__createBinding(exports, clerk_react_1, "useUser");
tslib_1.__createBinding(exports, clerk_react_1, "withClerk");
tslib_1.__createBinding(exports, clerk_react_1, "WithClerk");
tslib_1.__createBinding(exports, clerk_react_1, "withSession");
tslib_1.__createBinding(exports, clerk_react_1, "WithSession");
tslib_1.__createBinding(exports, clerk_react_1, "withUser");
tslib_1.__createBinding(exports, clerk_react_1, "WithUser");
tslib_1.__createBinding(exports, clerk_react_1, "isClerkAPIResponseError");
tslib_1.__createBinding(exports, clerk_react_1, "isMagicLinkError");
tslib_1.__createBinding(exports, clerk_react_1, "isMetamaskError");
tslib_1.__createBinding(exports, clerk_react_1, "isKnownError");
tslib_1.__createBinding(exports, clerk_react_1, "MagicLinkErrorCode");
var singleton_1 = require("./singleton");
tslib_1.__createBinding(exports, singleton_1, "clerk", "Clerk");
tslib_1.__exportStar(require("./ClerkProvider"), exports);
tslib_1.__exportStar(require("./useOAuth"), exports);
